nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java

```diff
@@ -0,0 +1,548 @@
+            if (cause == null) cause = se;
+            // This is a failure on the incoming data, rollback on failure if specified; otherwise route to failure after penalizing (and abort txn in any case)
+            if (rollbackOnFailure) {
+                abortConnection(hiveStreamingConnection);
+                throw new ProcessException(cause.getLocalizedMessage(), cause);
+            } else {
+                flowFile = session.penalize(flowFile);
+                Map<String, String> updateAttributes = new HashMap<>();
+                updateAttributes.put(HIVE_STREAMING_RECORD_COUNT_ATTR, Long.toString(hiveStreamingConnection.getConnectionStats().getRecordsWritten()));
+                updateAttributes.put(ATTR_OUTPUT_TABLES, options.getQualifiedTableName());
+                flowFile = session.putAllAttributes(flowFile, updateAttributes);
+                session.transfer(flowFile, REL_FAILURE);
+            }
+
+        } catch (Exception e) {
+            abortConnection(hiveStreamingConnection);
+            throw (e instanceof ProcessException) ? (ProcessException) e : new ProcessException(e);
+        } finally {
+            closeConnection(hiveStreamingConnection);
+            // Restore original class loader, might not be necessary but is good practice since the processor task changed it
+            Thread.currentThread().setContextClassLoader(originalClassloader);
+        }
+    }
+
+    StreamingConnection makeStreamingConnection(HiveOptions options, RecordReader reader) throws StreamingException {
+        return HiveStreamingConnection.newBuilder()
+                .withDatabase(options.getDatabaseName())
+                .withTable(options.getTableName())
+                .withStaticPartitionValues(options.getStaticPartitionValues())
+                .withHiveConf(options.getHiveConf())
+                .withRecordWriter(new HiveRecordWriter(reader, getLogger()))
+                .withAgentInfo("NiFi " + this.getClass().getSimpleName() + " [" + this.getIdentifier()
+                        + "] thread " + Thread.currentThread().getId() + "[" + Thread.currentThread().getName() + "]")
+                .connect();
+    }
+
+    @OnStopped
+    public void cleanup() {
+        validationResourceHolder.set(null); // trigger re-validation of resources
+
+        ComponentLog log = getLogger();
+
+        if (callTimeoutPool != null) {
+            callTimeoutPool.shutdown();
+            try {
+                while (!callTimeoutPool.isTerminated()) {
+                    callTimeoutPool.awaitTermination(callTimeout, TimeUnit.MILLISECONDS);
+                }
+            } catch (Throwable t) {
+                log.warn("shutdown interrupted on " + callTimeoutPool, t);
+            }
+            callTimeoutPool = null;
+        }
+
+        ugi = null;
+    }
+
+    private void abortAndCloseConnection(StreamingConnection connection) {
+        try {
+            abortConnection(connection);
```

Comment:
When does this get called? https://issues.apache.org/jira/browse/HIVE-19772 adds a shutdown handler and uncaught exception handler to auto close on interrupts or runtime exceptions. To be on safer side NiFi can also add interrupt handler which will invoke abortAndCloseConnection.

Reply:
Can you describe how to use the interrupt handler? This is left over from the Storm Hive bolt using the old API, only to abort the single transaction currently underway, if there was an error. Apache NiFi will be built on Apache Hive 3.0.x so we'd need HIVE-19772 to be in Hive 3.0.1 (and it be released) in order for us to take advantage of that. Otherwise we'd have to keep this manual abort in the meantime.

Reply:
Add Shutdown handler that invokes abortAndClose. Shutdown handler gets invoked on JVM shutdown. HIVE-19772 has example of shutdown handler and uncaught exception handler usage both of which can be added to Nifi processor. Basically the idea is to abort or commit properly when there is interrupt or runtime exception/errors. Without it, we may end up generating broken orc files which will not be cleaned up (compactor currently does not handle this gracefully)

Reply:
Will add the handler. However what happens when an error occurs during the writing of rows? Currently we catch various exceptions and if we're in a state where we shouldn't commit, we abort the transaction and send the flow file on (it can be retried in whole if the user routes it back to the input)